### PR TITLE
Add coroutine module

### LINKED ISSUE #30: coroutines
Hey. Do you have any plans to support coroutines?

### PR DESCRIPTION
**Pull request by matir91**
Fixes #30

### CODE REVIEW

File: fountain-retrofit/src/main/java/com/xmartlabs/fountain/retrofit/adapter/NetworkDataSourceExtensions.kt [hunk 1]
```diff
@@ -47,8 +43,12 @@ internal fun <T> Call<T>.doOnSuccess(onSuccessResponse: (T) -> Unit): Call<T> {
             callback?.onFailure(call, t) ?: Unit
 
         override fun onResponse(call: Call<T>?, response: Response<T>) {
-          response.body()?.let {
-            onSuccessResponse.invoke(it)
+          response.body().let {
+            if (it == null){
+              callback?.onFailure(call, IllegalStateException("Response cannot be null"))
+            }else {
```
Comment (FerrariBruno) [inline fountain-retrofit/src/main/java/com/xmartlabs/fountain/retrofit/adapter/NetworkDataSourceExtensions.kt]: Indentation

Comment (FerrariBruno) [reply fountain-retrofit/src/main/java/com/xmartlabs/fountain/retrofit/adapter/NetworkDataSourceExtensions.kt]: If you don't change this with the `orDo` extension

File: fountain-coroutines/src/main/java/com/xmartlabs/fountain/coroutines/adapter/NetworkDataSourceExtensions.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.xmartlabs.fountain.adapter.NetworkDataSourceAdapter
+import com.xmartlabs.fountain.adapter.NetworkResultListener
+import com.xmartlabs.fountain.adapter.PageFetcher
+import kotlinx.coroutines.experimental.runBlocking
+
+internal fun <T> CoroutinePageFetcher<T>.toPageFetcher(): PageFetcher<T> {
+  return object : PageFetcher<T> {
+    override fun fetchPage(page: Int, pageSize: Int, networkResultListener: NetworkResultListener<T>) {
+      runBlocking {
```
Comment (FerrariBruno) [inline fountain-coroutines/src/main/java/com/xmartlabs/fountain/coroutines/adapter/NetworkDataSourceExtensions.kt]: shouldn't this method be called on a `WorkerThread`?

File: fountain-rx2/src/main/java/com/xmartlabs/fountain/rx2/adapter/NetworkDataSourceExtensions.kt [hunk 3]
```diff
@@ -6,20 +6,14 @@ import com.xmartlabs.fountain.adapter.PageFetcher
...
-  override fun onSubscribe(d: Disposable) {}
-
-  override fun onError(e: Throwable) = this@toSingleObserver.onError(e)
-}
-
-internal fun <T> RxPageFetcher<T>.toPageFetcher()
-    : PageFetcher<T> {
+internal fun <T> RxPageFetcher<T>.toPageFetcher(): PageFetcher<T> {
   return object : PageFetcher<T> {
     override fun fetchPage(page: Int, pageSize: Int, networkResultListener: NetworkResultListener<T>) {
```
Comment (FerrariBruno) [inline fountain-rx2/src/main/java/com/xmartlabs/fountain/rx2/adapter/NetworkDataSourceExtensions.kt]: same?

File: fountain-coroutines/src/main/java/com/xmartlabs/fountain/coroutines/adapter/CoroutineNetworkDataSourceAdapterFactory.kt [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+   */
+  fun <Value, ListResponseValue : ListResponseWithEntityCount<Value>> fromTotalEntityCountListResponse(
+      pageFetcher: CoroutinePageFetcher<ListResponseValue>,
+      firstPage: Int = FountainConstants.DEFAULT_FIRST_PAGE
+  ) = object : CoroutineNetworkDataSourceAdapter<ListResponseValue> {
+    private val knownSizeResponseManager = KnownSizeResponseManager(firstPage)
+
+    override val coroutinePageFetcher: CoroutinePageFetcher<ListResponseValue>
+      get() = object : CoroutinePageFetcher<ListResponseValue> {
+        override fun fetchPage(page: Int, pageSize: Int): Deferred<ListResponseValue> {
```
Comment (FerrariBruno) [inline fountain-coroutines/src/main/java/com/xmartlabs/fountain/coroutines/adapter/CoroutineNetworkDataSourceAdapterFactory.kt]: can we annotate this with `@WorkerThread` somehow?

File: app/src/main/java/com/xmartlabs/sample/repository/UserRepositoryUsingCoroutines.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    })
+
+    val networkDataSourceAdapter = pageFetcher.toTotalEntityCountCoroutineNetworkDataSourceAdapter()
+    return FountainCoroutines.createNetworkListing(
+        networkDataSourceAdapter = networkDataSourceAdapter,
+        pagedListConfig = pagedListConfig
+    )
+  }
+
+  @MainThread
```
Comment (f7deleon) [inline app/src/main/java/com/xmartlabs/sample/repository/UserRepositoryUsingCoroutines.kt]: same

File: app/src/main/java/com/xmartlabs/sample/repository/UserRepositoryUsingRx.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class UserRepositoryUsingRx @Inject constructor(
+    private val userService: UserService,
+    private val userDao: UserDao,
+    private val db: AppDb
+) {
+  @MainThread
```
Comment (f7deleon) [inline app/src/main/java/com/xmartlabs/sample/repository/UserRepositoryUsingRx.kt]: same

File: app/src/main/java/com/xmartlabs/sample/repository/UserRepositoryUsingRx.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    })
+
+    val networkDataSourceAdapter = pageFetcher.toTotalEntityCountRetrofitNetworkDataSourceAdapter()
+    return FountainRx.createNetworkListing(
+        networkDataSourceAdapter = networkDataSourceAdapter,
+        pagedListConfig = pagedListConfig
+    )
+  }
+
+  @MainThread
```
Comment (f7deleon) [inline app/src/main/java/com/xmartlabs/sample/repository/UserRepositoryUsingRx.kt]: same

**Pull request merged**